streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java

```diff
@@ -109,16 +109,27 @@ public StateDirectory(final StreamsConfig config, final Time time, final boolean
                 log.warn("Using /tmp directory in the state.dir property can cause failures with writing the checkpoint file" +
                     " due to the fact that this directory can be cleared by the OS");
             }
-
             // change the dir permission to "rwxr-x---" to avoid world readable
-            final Path basePath = Paths.get(baseDir.getPath());
-            final Path statePath = Paths.get(stateDir.getPath());
+            configurePermissions(Paths.get(baseDir.getPath()));
+            configurePermissions(Paths.get(stateDir.getPath()));
+        }
+    }
+    
+    private void configurePermissions(final Path path) {
+        if (path.getFileSystem().supportedFileAttributeViews().contains("posix")) {
             final Set<PosixFilePermission> perms = PosixFilePermissions.fromString("rwxr-x---");
             try {
-                Files.setPosixFilePermissions(basePath, perms);
-                Files.setPosixFilePermissions(statePath, perms);
+                Files.setPosixFilePermissions(path, perms);
             } catch (final IOException e) {
-                log.error("Error changing permissions for the state or base directory {} ", stateDir.getPath(), e);
+                log.error("Error changing permissions for the directory {} ", path, e);
+            }
+        } else {
+            final File file = path.toFile();
```

Comment:
We pass the `path` parameter which is converted from `File` instance (i.e. `Paths.get(stateDir.getPath()`), and then we converted it back to `File` instance in the method. It's redundant. Could we just pass the `File` instance into the method directly?

Reply:
We'd have to pass in both the `Path` and the `File` and introduce the risk of them not being the "same". We need the `Path` to check if `PosixFilePermissions` can be set. It's purely stylistic, but I prefer it as currently written. Your call, though.

Reply:
I know we need both `File` and `Path` instance to complete the permission setting. What I meant is that:
1. We already had the `File` instances, i.e. baseDir, stateDir
2. `Paths.get(stateDir.getPath())`
--> what happened behind the scene, is we created a Path instance from the path string in the file instance (i.e.`stateDir.getPath()`)
```java
public final Path getPath(String var1, String... var2) {
... 
    return new UnixPath(this, var3);
}
```
3. `path.toFile()`
--> what happened behind the scene, is we created a File instance from the path string
```java
public final File toFile() {
        return new File(this.toString());
    }
```

So, currently, we will created an additional File instance via `path.toFile()`. If we pass the File instance into `configurePermissions()` directly, it'll only created a `Path` instance. ex: 
```java
configurePermissions(stateDir);  // pass the file instance directly
```
```java
private void configurePermissions(final File file) {
    final Path path = Paths.get(file.getPath()); // created path instance
    if (path.getFileSystem().supportedFileAttributeViews().contains("posix")) {
    ....
    }
    else {
            boolean set = file.setReadable(true, true);  // won't create additional File instance here
            ....
    }
}
```

Reply:
Sorry, I see what you mean now. I've pushed another commit that avoids the additional `File` creation.